drivers/crypto/crypto_rpi_pico_sha256.c

```diff
@@ -0,0 +1,202 @@
+/*
+ * Copyright (c) 2025 TOKITA Hiroshi
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT raspberrypi_pico_sha256
+
+#include <zephyr/crypto/crypto.h>
+
+#include <hardware/sha256.h>
+#include <pico/bootrom/lock.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(sha_rpi_pico_sha256, CONFIG_CRYPTO_LOG_LEVEL);
+
+/*
+ * This is based on the pico_sha256 implementation in the Pico-SDK.
+ */
+
+/* We add one 0x80 byte, then 8 bytes for the size */
+#define SHA256_PADDING_DATA_BYTES 9
+#define SHA256_BLOCK_SIZE_BYTES   64
+
+struct rpi_pico_sha256_data {
+	bool locked;
+	uint8_t cache_used;
+	union {
+		uint32_t word;
+		uint8_t bytes[4];
+	} cache;
+	size_t total_data_size;
+};
+
+static void write_to_hardware(struct rpi_pico_sha256_data *state, const uint8_t *data,
+			      size_t data_size_bytes)
+{
+	if (!state->cache_used && !(((uintptr_t)data) & 3u)) {
+		/* aligned writes */
+		while (data_size_bytes >= 4) {
+			/* write a whole word */
+			sha256_wait_ready_blocking();
+			sha256_put_word(*(const uint32_t *)data);
+			data += 4;
+			data_size_bytes -= 4;
+		}
+	}
+
+	while (data_size_bytes--) {
+		state->cache.bytes[state->cache_used++] = *data++;
+		if (state->cache_used == 4) {
+			state->cache_used = 0;
+			sha256_wait_ready_blocking();
+			sha256_put_word(state->cache.word);
+		}
+	}
+}
+
+static inline size_t bytes_to_write(struct rpi_pico_sha256_data *state, size_t data_size_bytes)
+{
+	return ((state->total_data_size + (SHA256_BLOCK_SIZE_BYTES - 1)) &
```

Comment:
I was wondering if here we could use `ROUND_UP()` instead of this (and similarly on line 99). Your solution seems to be more performant since it uses bit-masks, whereas the one in `zephyr/include/zephyr/sys/util.h` might rely on compiler to optimize that computation. OTOH it would be more readable. Wdyt?

Reply:
Thank you for your good pointing out.
I got enough optimized code with ROUND_UP().
I applied your suggestion.